searchlib/src/vespa/searchlib/query/tree/termnodes.cpp

```diff
@@ -1,26 +1,111 @@
 // Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
 
 #include "termnodes.h"
+#include <vespa/vespalib/util/exceptions.h>
+#include <charconv>
 
+using vespalib::IllegalArgumentException;
+using vespalib::stringref;
+using vespalib::make_string_short::fmt;
 namespace search::query {
 
-NumberTerm::~NumberTerm() = default;
-
-PrefixTerm::~PrefixTerm() = default;
-
-RangeTerm::~RangeTerm() = default;
-
 StringTerm::StringTerm(const Type &term, vespalib::stringref view, int32_t id, Weight weight)
     : QueryNodeMixinType(term, view, id, weight)
 {}
-StringTerm::~StringTerm() = default;
 
+NumberTerm::~NumberTerm() = default;
+PrefixTerm::~PrefixTerm() = default;
+RangeTerm::~RangeTerm() = default;
+StringTerm::~StringTerm() = default;
 SubstringTerm::~SubstringTerm() = default;
-
 SuffixTerm::~SuffixTerm() = default;
-
 LocationTerm::~LocationTerm() = default;
-
 RegExpTerm::~RegExpTerm() = default;
+WeightedSetTerm::~WeightedSetTerm() = default;
+DotProduct::~DotProduct() = default;
+WandTerm::~WandTerm() = default;
+
+namespace {
+
+void badType(const char *expected, const char *got) __attribute__((noinline));
+void badType(const char *expected, const char *got) {
+    throw IllegalArgumentException(fmt("Expected '%s' type, got '%s'", expected, got), VESPA_STRLOC);
+}
+
+class StringTermVector final : public MultiTerm::TermVector {
+public:
+    StringTermVector(uint32_t sz) : _terms() { _terms.reserve(sz); }
+    void addTerm(stringref term, Weight weight) override {
+        _terms.emplace_back(term, weight);
+    }
+    void addTerm(int64_t, Weight) override {
+        badType("string", "int64_t");
```

Comment:
this is a very well hidden restriction of the query stack format... how is this modeled in Java? is the term data type annotated on the parent multi-term node?

Reply:
Restriction lifted